File: internal/core/registry.go
package core

import "strings"

type ProviderID string
type RouteID string

type Money struct {
	Currency   string `json:"currency"`
	MinorUnits uint64 `json:"minor_units"`
}

func NewMoney(currency string, minorUnits uint64) Money {
	return Money{
		Currency:   strings.ToUpper(currency),
		MinorUnits: minorUnits,
	}
}

type AmountRange struct {
	MinMinorUnits uint64 `json:"min_minor_units"`
	MaxMinorUnits uint64 `json:"max_minor_units"`
}

func (r AmountRange) Contains(amount Money) bool {
	return amount.MinorUnits >= r.MinMinorUnits && amount.MinorUnits <= r.MaxMinorUnits
}

type Corridor struct {
	OriginCountry      string `json:"origin_country"`
	DestinationCountry string `json:"destination_country"`
}

func NewCorridor(originCountry, destinationCountry string) Corridor {
	return Corridor{
		OriginCountry:      strings.ToUpper(originCountry),
		DestinationCountry: strings.ToUpper(destinationCountry),
	}
}

type PayoutMethod string

const (
	PayoutBankAccount PayoutMethod = "bank_account"
	PayoutCashPickup  PayoutMethod = "cash_pickup"
	PayoutWallet      PayoutMethod = "wallet"
	PayoutCard        PayoutMethod = "card"
	PayoutBranch      PayoutMethod = "branch"
)

type RouteStatus string

const (
	RouteEnabled         RouteStatus = "enabled"
	RouteDisabled        RouteStatus = "disabled"
	RouteDegraded        RouteStatus = "degraded"
	RouteBlocked         RouteStatus = "blocked"
	RouteRecoveryTesting RouteStatus = "recovery_testing"
)

type Provider struct {
	ID       ProviderID `json:"id"`
	Name     string     `json:"name"`
	Enabled  bool       `json:"enabled"`
	Approved bool       `json:"approved"`
}

type Route struct {
	ID                 RouteID      `json:"id"`
	ProviderID         ProviderID   `json:"provider_id"`
	Corridor           Corridor     `json:"corridor"`
	PayoutMethod       PayoutMethod `json:"payout_method"`
	AmountRange        AmountRange  `json:"amount_range"`
	SettlementCurrency string       `json:"settlement_currency"`
	DestinationBanks   []string     `json:"destination_banks"`
	Status             RouteStatus  `json:"status"`
	LiquidityAvailable bool         `json:"liquidity_available"`
	FXRateAgeSeconds   *uint64      `json:"fx_rate_age_seconds,omitempty"`
	CostPenaltyBps     uint16       `json:"cost_penalty_bps"`
	FXQualityBps       uint16       `json:"fx_quality_bps"`
}

func (r Route) SupportsDestinationBank(destinationBank string) bool {
	if len(r.DestinationBanks) == 0 {
		return true
	}

	if destinationBank == "" {
		return false
	}

	normalized := strings.ToUpper(destinationBank)
	for _, bank := range r.DestinationBanks {
		if bank == normalized {
			return true
		}
	}
	return false
}

type RouteRegistry struct {
	Providers map[ProviderID]Provider `json:"providers"`
	Routes    map[RouteID]Route       `json:"routes"`
}

func NewRouteRegistry() RouteRegistry {
	return RouteRegistry{
		Providers: map[ProviderID]Provider{},
		Routes:    map[RouteID]Route{},
	}
}

func (r *RouteRegistry) AddProvider(provider Provider) {
	r.Providers[provider.ID] = provider
}

func (r *RouteRegistry) AddRoute(route Route) {
	r.Routes[route.ID] = route
}

func (r RouteRegistry) Provider(id ProviderID) (Provider, bool) {
	provider, ok := r.Providers[id]
	return provider, ok
}

func (r RouteRegistry) RoutesForCorridor(corridor Corridor) []Route {
	routes := make([]Route, 0)
	for _, route := range r.Routes {
		if route.Corridor == corridor {
			routes = append(routes, route)
		}
	}
	return routes
}

File: crates/imsi-core/src/registry.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub type ProviderId = String;
pub type RouteId = String;

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct Money {
    pub currency: String,
    pub minor_units: u64,
}

impl Money {
    pub fn new(currency: impl Into<String>, minor_units: u64) -> Self {
        Self {
            currency: currency.into().to_uppercase(),
            minor_units,
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct AmountRange {
    pub min_minor_units: u64,
    pub max_minor_units: u64,
}

impl AmountRange {
    pub fn contains(&self, amount: &Money) -> bool {
        amount.minor_units >= self.min_minor_units && amount.minor_units <= self.max_minor_units
    }
}

#[derive(Clone, Debug, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub struct Corridor {
    pub origin_country: String,
    pub destination_country: String,
}

impl Corridor {
    pub fn new(origin_country: impl Into<String>, destination_country: impl Into<String>) -> Self {
        Self {
            origin_country: origin_country.into().to_uppercase(),
            destination_country: destination_country.into().to_uppercase(),
        }
    }
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum PayoutMethod {
    BankAccount,
    CashPickup,
    Wallet,
    Card,
    Branch,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RouteStatus {
    Enabled,
    Disabled,
    Degraded,
    Blocked,
    RecoveryTesting,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct Provider {
    pub id: ProviderId,
    pub name: String,
    pub enabled: bool,
    pub approved: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct Route {
    pub id: RouteId,
    pub provider_id: ProviderId,
    pub corridor: Corridor,
    pub payout_method: PayoutMethod,
    pub amount_range: AmountRange,
    pub settlement_currency: String,
    pub destination_banks: Vec<String>,
    pub status: RouteStatus,
    pub liquidity_available: bool,
    pub fx_rate_age_seconds: Option<u64>,
    /// Lower is cheaper. Basis points above the best known route for the same corridor.
    pub cost_penalty_bps: u16,
    /// Higher is better. Includes rate quality and spread competitiveness.
    pub fx_quality_bps: u16,
}

impl Route {
    pub fn supports_destination_bank(&self, destination_bank: Option<&str>) -> bool {
        if self.destination_banks.is_empty() {
            return true;
        }

        destination_bank.is_some_and(|bank| {
            let bank = bank.to_uppercase();
            self.destination_banks.iter().any(|item| item == &bank)
        })
    }
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct RouteRegistry {
    pub providers: BTreeMap<ProviderId, Provider>,
    pub routes: BTreeMap<RouteId, Route>,
}

impl RouteRegistry {
    pub fn add_provider(&mut self, provider: Provider) {
        self.providers.insert(provider.id.clone(), provider);
    }

    pub fn add_route(&mut self, route: Route) {
        self.routes.insert(route.id.clone(), route);
    }

    pub fn provider(&self, provider_id: &str) -> Option<&Provider> {
        self.providers.get(provider_id)
    }

    pub fn routes_for_corridor<'a>(
        &'a self,
        corridor: &'a Corridor,
    ) -> impl Iterator<Item = &'a Route> + 'a {
        self.routes
            .values()
            .filter(move |route| &route.corridor == corridor)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_money_and_corridor_codes() {
        let money = Money::new("usd", 10_000);
        let corridor = Corridor::new("us", "ng");

        assert_eq!(money.currency, "USD");
        assert_eq!(corridor.origin_country, "US");
        assert_eq!(corridor.destination_country, "NG");
    }

    #[test]
    fn matches_destination_bank_when_limited() {
        let route = Route {
            id: "route-1".into(),
            provider_id: "provider-1".into(),
            corridor: Corridor::new("US", "NG"),
            payout_method: PayoutMethod::BankAccount,
            amount_range: AmountRange {
                min_minor_units: 1,
                max_minor_units: 1_000_000,
            },
            settlement_currency: "NGN".into(),
            destination_banks: vec!["ZENITH".into()],
            status: RouteStatus::Enabled,
            liquidity_available: true,
            fx_rate_age_seconds: Some(10),
            cost_penalty_bps: 0,
            fx_quality_bps: 9_000,
        };

        assert!(route.supports_destination_bank(Some("zenith")));
        assert!(!route.supports_destination_bank(Some("gtbank")));
        assert!(!route.supports_destination_bank(None));
    }
}
